File: Pow(x, n).java
class Solution {
    HashMap<Integer, Double> mp = new HashMap<Integer, Double>();
    public Double solve1(Double x, int n) {
        if (n == 0) {
            double one = 1 ;
            return one ;
        } else if (n == 1) {
            return x ;
        } else if (n % 2 == 0) {
            if (mp.get(n / 2) != null) {
                return mp.get(n / 2) * mp.get(n / 2) ;
            } else {
                mp.put(n / 2, solve1(x, n / 2)) ;
                return mp.get(n / 2) * mp.get(n / 2) ;
            }
        } else if (n % 2 != 0) {
            return solve1(x, n + 1) / x ;
        }
        return x ;
    }

    public Double solve2(Double x, int n) {
        if (n == 0) {
            double one = 1 ;
            return one ;
        } else if (n == 1) {
            return x ;
        } else if (n % 2 == 0) {
            if (mp.get(n / 2) != null) {
                return mp.get(n / 2) * mp.get(n / 2) ;
            } else {
                mp.put(n / 2, solve1(x, n / 2)) ;
                return mp.get(n / 2) * mp.get(n / 2) ;
            }
        } else if (n % 2 != 0) {
            return solve2(x, n - 1) * x ;
        }
        return x ;
    }

    public Double myPow(Double x, int n) {
        if (x == 0 && n == 0) {
            double one = 1 ;
            return one ;
        } else if (x == 0 || x == 1) {
            return x ;
        } else {
            if (n == 0) {
                double one = 1 ;
                return one ;
            } else if (n < 0) {
                return solve1(x, n) ;
            } else if (n > 0) {
                return solve2(x, n) ;
            }
        }
        return x ;
    }
}

File: Pow(x, n).cpp
class Solution {
public:
    unordered_map<int, double> mp ;
    double solve1(double x, int n) {
        if (n == 0) {
            return 1 ;
        }
        else if (n == 1) {
            return x ;
        }
        else if (n % 2 == 0) {
            if (mp[n / 2] != 0) {
                return mp[n / 2] * mp[n / 2] ;
            }
            else {
                mp[n / 2] = solve1(x, n / 2) ;
                return mp[n / 2] * mp[n / 2] ;
            }
        }
        else if (n % 2 != 0) {
            return solve1(x, n + 1) / x ;
        }
        return x ;
    }

    double solve2(double x, int n) {
        if (n == 0) {
            return 1 ;
        }
        else if (n == 1) {
            return x ;
        }
        else if (n % 2 == 0) {
            if (mp[n / 2] != 0) {
                return mp[n / 2] * mp[n / 2] ;
            }
            else {
                mp[n / 2] = solve1(x, n / 2) ;
                return mp[n / 2] * mp[n / 2] ;
            }
        }
        else if (n % 2 != 0) {
            return solve1(x, n - 1) * x ;
        }
        return x ;
    }

    double myPow(double x, int n) {
        if (x == 0 && n == 0) return 1 ;
        else if (x == 0 || x == 1) return x ;
        else {
            if (n == 0) {
                return 1 ;
            }
            else if (n < 0) {
                return solve1(x, n) ;
            }
            else if (n > 0) {
                return solve2(x, n) ;
            }
        }
        return x ;
    }
};